build_tools/scripts/compute_common_version.py

```diff
@@ -0,0 +1,63 @@
+# Copyright 2024 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# This scripts grabs the X.Y.Z[.dev]` version identifier from the
+# compiler's and runtime's `version.json` files and computes the
+# shared version.
+
+import argparse
+from pathlib import Path
+import json
+from datetime import datetime
+import subprocess
+
+from packaging.version import Version
+
+
+parser = argparse.ArgumentParser()
+release_type = parser.add_mutually_exclusive_group()
+release_type.add_argument("-rc", "--nightly-release", action="store_true")
+release_type.add_argument("-dev", "--development-release", action="store_true")
+release_type.add_argument("--custom-string", action="store", type=str)
+args = parser.parse_args()
+
+THIS_DIR = Path(__file__).parent.resolve()
+
+VERSION_FILE_COMPILER = THIS_DIR / "../../compiler/version.json"
+VERSION_FILE_RUNTIME = THIS_DIR / "../../runtime/version.json"
+
+
+def load_version_info(version_file):
+    with open(version_file, "rt") as f:
+        return json.load(f)
+
+
+compiler_version = load_version_info(VERSION_FILE_COMPILER)
+COMPILER_PACKAGE_VERSION = compiler_version.get("package-version")
+COMPILER_BASE_VERSION = Version(COMPILER_PACKAGE_VERSION).base_version
+
+runtime_version = load_version_info(VERSION_FILE_RUNTIME)
+RUNTIME_PACKAGE_VERSION = runtime_version.get("package-version")
+RUNTIME_BASE_VERSION = Version(RUNTIME_PACKAGE_VERSION).base_version
```

Comment:
Right now, the computed version is "returned" via `print(CURRENT_VERSION)` and the output of the script is captured in a bash variable
https://github.com/iree-org/iree/blob/ed4bc28802df90d5a76aecfac61052ce07f26720/.github/workflows/oneshot_candidate_release.yml#L21
This would break when adding logging via calling `print`. An alternative could be to pass the computed version via `sys.exit()` but I am not in favor of it.

Any suggestions?

Reply:
How about moving more of the scripting to a Python script, where you can freely log and write outputs all in a single process, without needing to juggle between bash and Python?

https://stackoverflow.com/a/70123641 has tips for writing to `$GITHUB_ENV` from Python

That script could detect if it is being run locally or on CI by the presence of that environment variable and log what it _would_ do on GitHub if you run it locally.

This could be a follow-up cleanup task, but I think it would help with future maintenance and debugging.

Reply:
Thanks for the suggestions. Based on those, I filled #19050 and will address it in a follow up.